File: app/frontend/src/pages/Kubernetes.js
import React, { useState, useEffect } from 'react';
import { kubernetesService } from '../services/api';
import DashboardCard from '../components/DashboardCard';
import InfoCard from '../components/InfoCard';
import TableComponent from '../components/TableComponent';
import StatusBadge from '../components/StatusBadge';
import LoadingSpinner from '../components/LoadingSpinner';
import { 
  RefreshCw, 
  Layers, 
  Cpu, 
  Network, 
  ShieldCheck, 
  HardDrive, 
  Grid,
  FileText,
  AlertCircle
} from 'lucide-react';
import './Kubernetes.css';

const Kubernetes = () => {
  const [data, setData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [refreshing, setRefreshing] = useState(false);
  const [error, setError] = useState('');
  const [activeTab, setActiveTab] = useState('pods');

  const fetchK8sData = async (showLoading = true) => {
    if (showLoading) setLoading(true);
    else setRefreshing(true);
    try {
      const response = await kubernetesService.getData();
      setData(response);
      setError('');
    } catch (err) {
      console.error(err);
      setError('Failed to fetch Kubernetes cluster resources. Please verify Flask backend.');
    } finally {
      setLoading(false);
      setRefreshing(false);
    }
  };

  useEffect(() => {
    fetchK8sData(true);
  }, []);

  if (loading) {
    return <LoadingSpinner message="Querying Kubernetes API server for cluster objects..." />;
  }

  if (error) {
    return (
      <div className="page-container">
        <div className="alert-error">
          <AlertCircle size={18} />
          <span>{error}</span>
        </div>
      </div>
    );
  }

  const { namespaces, pods, deployments, replica_sets, services, nodes, cluster_status } = data;

  const tabs = [
    { id: 'pods', label: 'Pods', count: pods.length },
    { id: 'deployments', label: 'Deployments', count: deployments.length },
    { id: 'replica_sets', label: 'Replica Sets', count: replica_sets.length },
    { id: 'services', label: 'Services', count: services.length },
    { id: 'namespaces', label: 'Namespaces', count: namespaces.length },
    { id: 'nodes', label: 'Nodes', count: nodes.length }
  ];

  // Helper renderers for each tab
  const renderTabContent = () => {
    switch (activeTab) {
      case 'pods':
        return (
          <TableComponent 
            headers={['Pod Name', 'Namespace', 'IP Address', 'Node', 'Age', 'Restarts', 'Status']}
            data={pods}
            renderRow={(pod, idx) => (
              <tr key={pod.name || idx} className="table-row">
                <td className="table-cell bold">{pod.name}</td>
                <td className="table-cell font-mono">{pod.namespace}</td>
                <td className="table-cell font-mono">{pod.ip}</td>
                <td className="table-cell">{pod.node}</td>
                <td className="table-cell">{pod.age}</td>
                <td className="table-cell text-center font-mono">{pod.restarts}</td>
                <td className="table-cell">
                  <StatusBadge status={pod.status} />
                </td>
              </tr>
            )}
          />
        );
      case 'deployments':
        return (
          <TableComponent
            headers={['Deployment Name', 'Namespace', 'Ready Pods', 'Up to Date', 'Available', 'Age']}
            data={deployments}
            renderRow={(dep, idx) => (
              <tr key={dep.name || idx} className="table-row">
                <td className="table-cell bold">{dep.name}</td>
                <td className="table-cell font-mono">{dep.namespace}</td>
                <td className="table-cell text-center font-mono bold">{dep.ready}</td>
                <td className="table-cell text-center font-mono">{dep.up_to_date}</td>
                <td className="table-cell text-center font-mono">{dep.available}</td>
                <td className="table-cell">{dep.age}</td>
              </tr>
            )}
          />
        );
      case 'replica_sets':
        return (
          <TableComponent
            headers={['ReplicaSet Name', 'Namespace', 'Desired', 'Current', 'Ready', 'Age']}
            data={replica_sets}
            renderRow={(rs, idx) => (
              <tr key={rs.name || idx} className="table-row">
                <td className="table-cell bold">{rs.name}</td>
                <td className="table-cell font-mono">{rs.namespace}</td>
                <td className="table-cell text-center font-mono">{rs.desired}</td>
                <td className="table-cell text-center font-mono">{rs.current}</td>
                <td className="table-cell text-center font-mono bold">{rs.ready}</td>
                <td className="table-cell">{rs.age}</td>
              </tr>
            )}
          />
        );
      case 'services':
        return (
          <TableComponent
            headers={['Service Name', 'Namespace', 'Type', 'Cluster IP', 'External IP', 'Ports', 'Age']}
            data={services}
            renderRow={(svc, idx) => (
              <tr key={svc.name || idx} className="table-row">
                <td className="table-cell bold">{svc.name}</td>
                <td className="table-cell font-mono">{svc.namespace}</td>
                <td className="table-cell text-success font-semibold">{svc.type}</td>
                <td className="table-cell font-mono">{svc.cluster_ip}</td>
                <td className="table-cell font-mono">{svc.external_ip}</td>
                <td className="table-cell font-mono">{svc.ports}</td>
                <td className="table-cell">{svc.age}</td>
              </tr>
            )}
          />
        );
      case 'namespaces':
        return (
          <TableComponent
            headers={['Namespace Name', 'Age', 'Status']}
            data={namespaces}
            renderRow={(ns, idx) => (
              <tr key={ns.name || idx} className="table-row">
                <td className="table-cell bold">{ns.name}</td>
                <td className="table-cell">{ns.age}</td>
                <td className="table-cell">
                  <StatusBadge status={ns.status} />
                </td>
              </tr>
            )}
          />
        );
      case 'nodes':
        return (
          <TableComponent
            headers={['Node Name', 'Role', 'IP Address', 'Kubelet Version', 'Age', 'Status']}
            data={nodes}
            renderRow={(node, idx) => (
              <tr key={node.name || idx} className="table-row">
                <td className="table-cell bold">{node.name}</td>
                <td className="table-cell text-info font-semibold">{node.roles}</td>
                <td className="table-cell font-mono">{node.internal_ip}</td>
                <td className="table-cell font-mono">{node.version}</td>
                <td className="table-cell">{node.age}</td>
                <td className="table-cell">
                  <StatusBadge status={node.status} />
                </td>
              </tr>
            )}
          />
        );
      default:
        return null;
    }
  };

  return (
    <div className="page-container">
      <div className="page-header">
        <div>
          <h1 className="page-title">Kubernetes Cluster Explorer</h1>
          <p className="page-subtitle">Inspect namespaces, nodes, deployments, and running pod configurations on the control plane.</p>
        </div>
        <button 
          onClick={() => fetchK8sData(false)} 
          className="btn-refresh" 
          disabled={refreshing}
        >
          <RefreshCw size={14} className={refreshing ? 'spin' : ''} />
          <span>Sync API Objects</span>
        </button>
      </div>

      {/* Cluster Status Summary cards */}
      <div className="k8s-summary-grid">
        <InfoCard label="API Server" value={cluster_status.api_server} icon={ShieldCheck} />
        <InfoCard label="Scheduler" value={cluster_status.scheduler} icon={Grid} />
        <InfoCard label="Controller Manager" value={cluster_status.controller_manager} icon={Cpu} />
        <InfoCard label="Cluster Health" value={cluster_status.status} icon={Layers} />
      </div>

      {/* Tab bar navigation */}
      <div className="k8s-tabs-container">
        <div className="tabs-bar">
          {tabs.map((tab) => (
            <button
              key={tab.id}
              onClick={() => setActiveTab(tab.id)}
              className={`tab-btn ${activeTab === tab.id ? 'active' : ''}`}
            >
              <span>{tab.label}</span>
              <span className="tab-count">{tab.count}</span>
            </button>
          ))}
        </div>

        {/* Tab content panel */}
        <div className="tab-content-panel">
          <DashboardCard 
            title={`${activeTab.replace('_', ' ').toUpperCase()} RESOURCE GRID`}
            subtitle={`Live telemetry queries from Kubernetes controller-manager namespaces.`}
          >
            {renderTabContent()}
          </DashboardCard>
        </div>
      </div>
    </div>
  );
};

export default Kubernetes;

File: app/backend/routes/kubernetes.py
from flask import Blueprint, jsonify

kubernetes_bp = Blueprint('kubernetes', __name__)

# Mock data reflecting a typical small-medium enterprise K8s cluster
mock_namespaces = [
    {"name": "default", "status": "Active", "age": "45d"},
    {"name": "kube-system", "status": "Active", "age": "45d"},
    {"name": "infrapilot-prod", "status": "Active", "age": "12d"},
    {"name": "monitoring", "status": "Active", "age": "20d"}
]

mock_pods = [
    {"name": "infrapilot-backend-7fd6b88989-2klm8", "namespace": "infrapilot-prod", "status": "Running", "restarts": 0, "ip": "10.244.1.42", "node": "node-worker-1", "age": "5d"},
    {"name": "infrapilot-backend-7fd6b88989-pxz9q", "namespace": "infrapilot-prod", "status": "Running", "restarts": 1, "ip": "10.244.2.19", "node": "node-worker-2", "age": "5d"},
    {"name": "infrapilot-frontend-6b44ccdf96-jhd78", "namespace": "infrapilot-prod", "status": "Running", "restarts": 0, "ip": "10.244.1.43", "node": "node-worker-1", "age": "5d"},
    {"name": "prometheus-server-85dfc4c79d-q8w2r", "namespace": "monitoring", "status": "Running", "restarts": 2, "ip": "10.244.2.20", "node": "node-worker-2", "age": "10d"},
    {"name": "grafana-7bb5df756b-pqr72", "namespace": "monitoring", "status": "Running", "restarts": 0, "ip": "10.244.1.44", "node": "node-worker-1", "age": "10d"},
    {"name": "postgres-db-0", "namespace": "default", "status": "Running", "restarts": 0, "ip": "10.244.2.21", "node": "node-worker-2", "age": "30d"}
]

mock_deployments = [
    {"name": "infrapilot-backend", "namespace": "infrapilot-prod", "ready": "2/2", "up_to_date": 2, "available": 2, "age": "12d"},
    {"name": "infrapilot-frontend", "namespace": "infrapilot-prod", "ready": "1/1", "up_to_date": 1, "available": 1, "age": "12d"},
    {"name": "prometheus-server", "namespace": "monitoring", "ready": "1/1", "up_to_date": 1, "available": 1, "age": "20d"},
    {"name": "grafana", "namespace": "monitoring", "ready": "1/1", "up_to_date": 1, "available": 1, "age": "20d"}
]

mock_replica_sets = [
    {"name": "infrapilot-backend-7fd6b88989", "namespace": "infrapilot-prod", "desired": 2, "current": 2, "ready": 2, "age": "5d"},
    {"name": "infrapilot-frontend-6b44ccdf96", "namespace": "infrapilot-prod", "desired": 1, "current": 1, "ready": 1, "age": "5d"},
    {"name": "prometheus-server-85dfc4c79d", "namespace": "monitoring", "desired": 1, "current": 1, "ready": 1, "age": "10d"},
    {"name": "grafana-7bb5df756b", "namespace": "monitoring", "desired": 1, "current": 1, "ready": 1, "age": "10d"}
]

mock_services = [
    {"name": "infrapilot-backend-svc", "namespace": "infrapilot-prod", "type": "ClusterIP", "cluster_ip": "10.96.42.11", "external_ip": "None", "ports": "5000/TCP", "age": "12d"},
    {"name": "infrapilot-frontend-svc", "namespace": "infrapilot-prod", "type": "LoadBalancer", "cluster_ip": "10.96.85.220", "external_ip": "192.168.1.150", "ports": "80:32080/TCP", "age": "12d"},
    {"name": "prometheus-service", "namespace": "monitoring", "type": "ClusterIP", "cluster_ip": "10.96.112.56", "external_ip": "None", "ports": "9090/TCP", "age": "20d"},
    {"name": "grafana-service", "namespace": "monitoring", "type": "NodePort", "cluster_ip": "10.96.201.89", "external_ip": "None", "ports": "80:31200/TCP", "age": "20d"}
]

mock_nodes = [
    {"name": "node-master", "status": "Ready", "roles": "control-plane", "age": "45d", "version": "v1.28.2", "internal_ip": "192.168.1.50"},
    {"name": "node-worker-1", "status": "Ready", "roles": "worker", "age": "45d", "version": "v1.28.2", "internal_ip": "192.168.1.51"},
    {"name": "node-worker-2", "status": "Ready", "roles": "worker", "age": "45d", "version": "v1.28.2", "internal_ip": "192.168.1.52"}
]

@kubernetes_bp.route('/kubernetes', methods=['GET'])
def get_kubernetes_data():
    """Retrieve all Kubernetes cluster object data and statuses."""
    cluster_status = {
        "status": "Healthy",
        "api_server": "Healthy",
        "scheduler": "Healthy",
        "controller_manager": "Healthy",
        "node_count": len(mock_nodes),
        "total_namespaces": len(mock_namespaces),
        "total_pods": len(mock_pods),
        "total_deployments": len(mock_deployments)
    }
    
    return jsonify({
        "namespaces": mock_namespaces,
        "pods": mock_pods,
        "deployments": mock_deployments,
        "replica_sets": mock_replica_sets,
        "services": mock_services,
        "nodes": mock_nodes,
        "cluster_status": cluster_status
    }), 200
